File: internal/collector/thermal.go
package collector

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/jefferymvp/openwrt_status/internal/model"
)

// GetThermalStatus 获取系统所有温度传感器的温度数据
func GetThermalStatus() model.ThermalStatus {
	sensors := make([]model.ThermalItem, 0)

	// 1. 读取 /sys/class/thermal/thermal_zone*
	tzSensors := readThermalZones()
	sensors = append(sensors, tzSensors...)

	// 2. 读取 /sys/class/hwmon/hwmon*
	hwmonSensors := readHwmon()
	sensors = append(sensors, hwmonSensors...)

	return model.ThermalStatus{
		Sensors: sensors,
	}
}

// readThermalZones 从 /sys/class/thermal/thermal_zone* 读取温度
func readThermalZones() []model.ThermalItem {
	zones, err := filepath.Glob("/sys/class/thermal/thermal_zone*")
	if err != nil || len(zones) == 0 {
		return nil
	}

	var results []model.ThermalItem
	for _, zoneDir := range zones {
		zoneName := filepath.Base(zoneDir)

		// 读取传感器类型名（如 cpu-thermal、soc-thermal）
		typeName := readStringFile(filepath.Join(zoneDir, "type"))
		if typeName == "" {
			typeName = zoneName
		}

		rawTemp := readFirstIntFromFile(filepath.Join(zoneDir, "temp"))
		if rawTemp <= 0 {
			continue
		}

		celsius := normalizeTemperature(rawTemp)
		results = append(results, model.ThermalItem{
			Name:        fmt.Sprintf("%s (%s)", typeName, zoneName),
			Temperature: celsius,
			Type:        "thermal_zone",
		})
	}

	return results
}

// readHwmon 从 /sys/class/hwmon/hwmon* 读取硬件传感器温度
func readHwmon() []model.ThermalItem {
	hwmons, err := filepath.Glob("/sys/class/hwmon/hwmon*")
	if err != nil || len(hwmons) == 0 {
		return nil
	}

	var results []model.ThermalItem
	for _, hwmonDir := range hwmons {
		hwmonName := readStringFile(filepath.Join(hwmonDir, "name"))
		if hwmonName == "" {
			hwmonName = filepath.Base(hwmonDir)
		}

		tempInputs, _ := filepath.Glob(filepath.Join(hwmonDir, "temp*_input"))
		for _, tempFile := range tempInputs {
			base := filepath.Base(tempFile) // 例如 temp1_input
			idx := strings.TrimSuffix(strings.TrimPrefix(base, "temp"), "_input")

			// 读取 label 文件（如 temp1_label）
			label := readStringFile(filepath.Join(hwmonDir, fmt.Sprintf("temp%s_label", idx)))
			if label == "" {
				label = fmt.Sprintf("Sensor %s", idx)
			}

			rawTemp := readFirstIntFromFile(tempFile)
			if rawTemp <= 0 {
				continue
			}

			celsius := normalizeTemperature(rawTemp)
			results = append(results, model.ThermalItem{
				Name:        fmt.Sprintf("%s - %s", hwmonName, label),
				Temperature: celsius,
				Type:        "hwmon",
			})
		}
	}

	return results
}

// normalizeTemperature 标准化温度值为摄氏度
func normalizeTemperature(raw int64) float64 {
	// Linux 内核驱动常规以毫摄氏度 (m°C) 呈现，如 45000 代表 45°C
	if raw > 1000 {
		return roundFloat(float64(raw)/1000.0, 1)
	}
	return roundFloat(float64(raw), 1)
}

// readStringFile 读取单行文本文件内容
func readStringFile(path string) string {
	content, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(content))
}

File: src/collector/thermal.rs
use std::fs;
use std::path::Path;

use super::round_float;
use crate::model::{ThermalItem, ThermalStatus};

/// 获取系统所有温度传感器的温度数据
pub fn get_thermal_status() -> ThermalStatus {
    let mut sensors = Vec::new();

    // 1. 读取 /sys/class/thermal/thermal_zone*
    sensors.extend(read_thermal_zones());

    // 2. 读取 /sys/class/hwmon/hwmon*
    sensors.extend(read_hwmon());

    ThermalStatus { sensors }
}

fn read_thermal_zones() -> Vec<ThermalItem> {
    let mut results = Vec::new();
    let base_path = Path::new("/sys/class/thermal");
    if !base_path.exists() {
        return results;
    }

    let entries = match fs::read_dir(base_path) {
        Ok(e) => e,
        Err(_) => return results,
    };

    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        if !name.starts_with("thermal_zone") {
            continue;
        }

        let type_path = entry.path().join("type");
        let temp_path = entry.path().join("temp");

        let type_name = fs::read_to_string(&type_path)
            .map(|s| s.trim().to_string())
            .unwrap_or_else(|_| name.clone());

        if let Ok(temp_str) = fs::read_to_string(&temp_path) {
            if let Ok(raw_temp) = temp_str.trim().parse::<i64>() {
                if raw_temp > 0 {
                    results.push(ThermalItem {
                        name: format!("{} ({})", type_name, name),
                        temperature: normalize_temperature(raw_temp),
                        r#type: Some("thermal_zone".to_string()),
                    });
                }
            }
        }
    }

    results
}

fn read_hwmon() -> Vec<ThermalItem> {
    let mut results = Vec::new();
    let base_path = Path::new("/sys/class/hwmon");
    if !base_path.exists() {
        return results;
    }

    let entries = match fs::read_dir(base_path) {
        Ok(e) => e,
        Err(_) => return results,
    };

    for entry in entries.flatten() {
        let hwmon_dir_name = entry.file_name().to_string_lossy().to_string();
        if !hwmon_dir_name.starts_with("hwmon") {
            continue;
        }

        let hwmon_name = fs::read_to_string(entry.path().join("name"))
            .map(|s| s.trim().to_string())
            .unwrap_or_else(|_| hwmon_dir_name.clone());

        let sub_entries = match fs::read_dir(entry.path()) {
            Ok(se) => se,
            Err(_) => continue,
        };

        for sub in sub_entries.flatten() {
            let sub_name = sub.file_name().to_string_lossy().to_string();
            // 匹配类似 temp1_input, temp2_input
            if sub_name.starts_with("temp") && sub_name.ends_with("_input") {
                let idx = sub_name
                    .trim_start_matches("temp")
                    .trim_end_matches("_input");
                let label_path = entry.path().join(format!("temp{}_label", idx));
                let label = fs::read_to_string(&label_path)
                    .map(|s| s.trim().to_string())
                    .unwrap_or_else(|_| format!("Sensor {}", idx));

                if let Ok(temp_str) = fs::read_to_string(sub.path()) {
                    if let Ok(raw_temp) = temp_str.trim().parse::<i64>() {
                        if raw_temp > 0 {
                            results.push(ThermalItem {
                                name: format!("{} - {}", hwmon_name, label),
                                temperature: normalize_temperature(raw_temp),
                                r#type: Some("hwmon".to_string()),
                            });
                        }
                    }
                }
            }
        }
    }

    results
}

fn normalize_temperature(raw: i64) -> f64 {
    if raw > 1000 {
        round_float(raw as f64 / 1000.0, 1)
    } else {
        round_float(raw as f64, 1)
    }
}
